<a href="https://colab.research.google.com/github/Lonebaithei/EDA_Assignment_2026/blob/main/T05_Continuous_Probability_Distributions_and_Statistical_Inference.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

SECTION B: Advanced Application — Instacart Dataset

In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
import os

from scipy.stats import skew, kurtosis
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


B1. Data Loading and Integrity Checks

In [2]:
orders = pd.read_csv('/content/drive/MyDrive/Colab_Notebooks/ADA2026_Instacart_Project/data/raw/Instacart/archive(1)/orders.csv')
products = pd.read_csv('/content/drive/MyDrive/Colab_Notebooks/ADA2026_Instacart_Project/data/raw/Instacart/archive(1)/products.csv')
order_products = pd.read_csv('/content/drive/MyDrive/Colab_Notebooks/ADA2026_Instacart_Project/data/raw/Instacart/archive(1)/order_products__prior.csv')

Tasks:

*   Identify primary keys.
*   Verify uniqueness of order id.
*   Investigate missing values in days since prior order.


**B2. Feature Engineering: Order Size**

In [3]:
ordersize = order_products.groupby('order_id').size ()


*   Compute mean and standard deviation.
*   Plot histogram.
*   Assess normality.


Computation of the **Mean** and **Standard Deviation**.

In [4]:
mean_ordersize = ordersize.mean()
std_ordersize = ordersize.std()

print(f"Mean order size: {mean_ordersize:.2f}")
print(f"Standard deviation of order size: {std_ordersize:.2f}")

Mean order size: 10.09
Standard deviation of order size: 7.53


Plotting a **Histogram** showing the distribution of **order_size**.

In [ ]:
plt.figure(figsize=(10, 6))
plt.hist(ordersize, bins=50, edgecolor='black', alpha=0.7)
plt.title('Distribution of Order Size')
plt.xlabel('Number of Products in an Order')
plt.ylabel('Frequency')
plt.grid(axis='y', alpha=0.75)
plt.show()

**Interpretation**

The mean order size is approximately 10.09, and the standard deviation is approximately 7.53. **This indicates that on average, an order contains about 10 products, with a moderate spread around this average**.

### Assessing Normality: Skewness and Kurtosis

In [ ]:
skewness = skew(ordersize)
kurt = kurtosis(ordersize)

print(f"Skewness of order size: {skewness:.2f}")
print(f"Kurtosis of order size: {kurt:.2f}")

if abs(skewness) < 0.5 and abs(kurt) < 0.5:
    print("The distribution appears approximately normal based on skewness and kurtosis.")
elif skewness > 0.5:
    print("The distribution is positively skewed (has a long tail to the right).")
elif skewness < -0.5:
    print("The distribution is negatively skewed (has a long tail to the left).")

if kurt > 0.5:
    print("The distribution is leptokurtic (has heavier tails and a sharper peak than a normal distribution).")
elif kurt < -0.5:
    print("The distribution is platykurtic (has lighter tails and a flatter peak than a normal distribution).")

### Assessing Normality: Q-Q Plot

In [ ]:
sm.qqplot(ordersize, line='s')
plt.title('Q-Q Plot of Order Size')
plt.show()

**B3. Z-Score Analysis**

*   Identify observations with |Z| ≥ 3.
*   Interpret unusual values.

In [ ]:
z_scores = (ordersize - mean_ordersize) / std_ordersize

outliers = ordersize[abs(z_scores) >= 3]

print(f"Number of observations with |Z| >= 3: {len(outliers)}")
print("Observations with |Z| >= 3 (first 10 if many):")
print(outliers.head(10))

**Interpretation of unusual values**

*   The Z-score analysis has been completed and we found 48,066 observations where the absolute Z-score was greater than or equal to 3.

*   These observations represent orders with an unusually high or low number of products compared to the average.
*   The output shows the order_id and the ordersize (number of products) for the first 10 of these potential outliers. For instance, order_id 106 had 34 products, order_id 742 had 45 products, which are significantly larger than the mean order size of approximately 10 products.

*    This indicates that these orders are quite large and deviate substantially from the typical order size.

**B4. CLT Simulation**

*   Simulate sampling distributions for:

*n* = 10, *n* = 30, *n* = 100


*   Comparing the **Shape** and **Variance**.